lib/posix/options/pthread.c

```diff
@@ -1115,8 +1111,18 @@ int pthread_join(pthread_t pthread, void **status)
 		break;
 	}
 
-	ret = k_thread_join(&t->thread, K_FOREVER);
-	/* other possibilities? */
+	ret = k_thread_join(&t->thread, timeout);
+	if (ret != 0) {
+		SYS_SEM_LOCK(&pthread_pool_lock) {
+			t->attr.detachstate = PTHREAD_CREATE_JOINABLE;
```

Comment:
I don't get this, I think a comment here would be nice

Reply:
Me neither. This the path @cfriedt send. Maybe, could you provide an explanatory comment?

Reply:
[Comment](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/posix/options/pthread.c#L1099) just a few lines above this in `main`.

Reply:
Only one thread can attempt to join another thread at a given time, so by changing the detached state to detached before calling `k_thread_join()` we can ensure that no other (p)threads could join it. However, if joining fails due to a timeout, we need to be able to retry later.

Reply:
I have added a comment.